app/services/experiment.py

```diff
@@ -0,0 +1,269 @@
+from app.models.experiment_run import ExperimentRun
+from app.models.experiment_template import ExperimentTemplate
+from app.schemas.experiment_run import ExperimentRunId
+from app.schemas.experiment_template import ExperimentTemplateId
+from app.schemas.states import RunState, TemplateState
+from app.services.workflow import ReanaService
+
+
+class ExperimentService:
+    DOCKER_SERVICE: ExperimentService | None = None
+
+    def __init__(self) -> None:
+        self.docker_client = DockerClient(base_url=settings.DOCKER_BASE_URL)
+        self.logger = logging.getLogger("uvicorn")
+
+        self.experiment_run_queue = asyncio.Queue()
+        self.image_building_queue = asyncio.Queue()
+
+        if not self.docker_registry_login():
+            raise SystemExit(
+                f"Unable to log in to Docker registry '{settings.DOCKER_REGISTRY_URL}'. Exiting..."
+            )
+
+        if not ReanaService.has_access():
+            raise SystemExit(
+                f"Unable to connect to REANA server '{settings.REANA_SERVER_URL}'. Exiting..."
+            )
+
+    def docker_registry_login(self) -> bool:
+        try:
+            response = self.docker_client.login(
+                username=settings.DOCKER_REGISTRY_USERNAME,
+                password=settings.DOCKER_REGISTRY_PASSWORD,
+                registry=settings.DOCKER_REGISTRY_URL,
+            )
+            if response["Status"] == "Login Succeeded":
+                return True
+        except Exception:
+            # TODO: Handle exception properly
+            self.logger.exception("Error occurred while connecting to Docker registry")
+
+        return False
+
+    async def init_run_queue(self) -> None:
+        run_ids = (
+            await ExperimentRun.find(
+                ExperimentRun.state != RunState.FINISHED,
+                ExperimentRun.state != RunState.CRASHED,
+            )
+            .sort(+ExperimentRun.updated_at)
+            .project(ExperimentRunId)
+            .to_list()
+        )
+        for run_id in run_ids:
+            await self.add_run_to_execute(run_id.id)
+            workflow_name = f"{EXPERIMENT_RUN_DIR_PREFIX}{str(run_id.id)}"
+            try:
+                ReanaService.stop_workflow(workflow_name, force_stop=True)
+                ReanaService.delete_workflow(
+                    workflow_name, all_runs=True, workspace=True
+                )
+            except Exception as e:
+                self.logger.warning(
+                    f"REANA Workflow of ExperimentRun id={run_id.id} "
+                    + "was not successfully deleted",
+                    exc_info=e,
+                )
+
+        count = self.experiment_run_queue.qsize()
+        if count > 0:
+            self.logger.info(
+                "Workflow queue has been initialized with "
+                + f"{count} workflows to execute"
+            )
+
+    async def init_image_build_queue(self) -> None:
+        run_ids = (
```

Comment:
`run_ids` should be renamed to `templated_ids` since they represent exp templates, not runs